### PR TITLE
Issue 761/assets by chunk name stats component json content hash

### PR DESCRIPTION
**Pull request by ComLock**

### CODE REVIEW

File: src/main/resources/lib/enonic/react4xp/asset/getImmuteableDependencies.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import {includes} from '@enonic/js-utils/array/includes';
+import endsWith from '@enonic/js-utils/string/endsWith';
+import { startsWith } from '@enonic/js-utils/string/startsWith';
+// import {toStr} from '@enonic/js-utils/value/toStr';
+import {getComponentStats} from './getComponentStats';
+
+
+export function getImmuteableDependencies(entries: string[]) {
```
Comment (alansemenov) [inline src/main/resources/lib/enonic/react4xp/asset/getImmuteableDependencies.ts:8]:
```suggestion
export function getImmutableDependencies(entries: string[]) {
```

File: src/main/resources/lib/enonic/react4xp/asset/getImmuteables.ts [hunk 2]
```diff
@@ -1,28 +1,41 @@
...
+
 	const clientChunkFilename = getClientChunkFilename();
+	// log.debug('getImmuteables clientChunkFilename:%s', toStr(clientChunkFilename)); // client-47ERBIZ7.global.js
+
 	const executorChunkFilename = getExecutorChunkFilename();
-	//log.debug('handleAssetRequest clientChunkFilename:%s', toStr(clientChunkFilename));
-	//log.debug('handleAssetRequest externals:%s', toStr(externals));
+	// log.debug('getImmuteables executorChunkFilename:%s', toStr(executorChunkFilename)); // executor-O7SLEQAY.js
 
 	const immuteables = {};
```
Comment (alansemenov) [inline src/main/resources/lib/enonic/react4xp/asset/getImmuteables.ts:21]:
```suggestion
	const immutables = {};
```

File: src/main/resources/lib/enonic/react4xp/htmlHandling.ts [hunk 3]
```diff
@@ -64,14 +64,23 @@ export function buildErrorContainer(
...
+// Content Security Policy directive: "script-src 'self'".
+// Either the 'unsafe-inline' keyword,
+// a hash ('sha256-1bp2x20QuWwY4nr/H7kToNvRaAdW2Dfg3fghhc9NzBk='),
+// or a nonce ('nonce-...') is required to enable inline execution.
+//
+// <script>console.error("${heading}${msg ? `:\\n${msg}` : ''}${
+// 	react4xpObj
+// 		? '\\n\\nJsxPath: ' + jsxPath + "\\nID: " + react4xpId
+// 		: ''
+// }\\n\\n${protip}");</script>
```
Comment (alansemenov) [inline src/main/resources/lib/enonic/react4xp/htmlHandling.ts:83]: Can we just remove it (also the comment)?

**Review by alansemenov — CHANGES_REQUESTED** [review]

File: src/main/resources/lib/enonic/react4xp/asset/getImmuteables.ts [hunk 4]
```diff
@@ -1,28 +1,41 @@
...
+
 	for (let i = 0; i < dependencies.length; i++) {
 		const dependency = dependencies[i];
 		immuteables[dependency] = true;
 	}
+
 	for (let i = 0; i < externals.length; i++) {
 		const external = externals[i];
-		immuteables[external] = true;
+		if (external !== 'externals.js') {
```
[the remark this answers is not in the document]
Comment (ComLock) [reply src/main/resources/lib/enonic/react4xp/asset/getImmuteables.ts:35]: For now yes. This filename will only change if we change it in the build system.
When building an app with the lib, one MUST use the correct version of the build system.

File: src/main/resources/lib/enonic/react4xp/htmlHandling.ts (hunk 3 above)
Comment (ComLock) [reply src/main/resources/lib/enonic/react4xp/htmlHandling.ts]: done

File: src/main/resources/lib/enonic/react4xp/asset/getImmuteables.ts (hunk 2 above)
Comment (ComLock) [reply src/main/resources/lib/enonic/react4xp/asset/getImmuteables.ts]: fixed

File: src/main/resources/lib/enonic/react4xp/asset/getImmuteableDependencies.ts (hunk 1 above)
Comment (ComLock) [reply src/main/resources/lib/enonic/react4xp/asset/getImmuteableDependencies.ts]: fixed

**Review by alansemenov — APPROVED** [review]

**Pull request merged**